### PR TITLE
✨ Add ERC6909 Minimal Multi-Token

### PR DESCRIPTION
**Pull request by Atarpara**
## Description

Added [erc6909](https://eips.ethereum.org/EIPS/eip-6909) implementation.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge fmt`?
- [x] Ran `forge snapshot`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

**Comment by Vectorized** [comment]
Oooooooooooooo!!!

**Comment by Atarpara** [comment]
@Vectorized this is ready to your review.

**Comment by jtriley-eth** [comment]
lgtm; very nice!

**Review by Vectorized — APPROVED** [review]
Imma merge into my own branch and fix from there. Thanks for the good work!

File: src/tokens/ERC6909.sol [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+    function allowance(address owner, address spender, uint256 id)
+        public
+        view
+        virtual
+        returns (uint256 amount)
+    {
+        /// @solidity memory-safe-assembly
+        assembly {
+            /// Cache the free memory pointer.
+            let m := mload(0x40)
```
[the remark this answers is not in the document]
Comment (Atarpara) [reply src/tokens/ERC6909.sol:165]: Oh. Good catch.

File: src/tokens/ERC6909.sol [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+            /// Compute the balance slot and load its value.
+            mstore(0x20, _ERC6909_MASTER_SLOT_SEED)
+            mstore(0x14, caller())
+            mstore(0x00, id)
+            let fromBalanceSlot := keccak256(0x00, 0x40)
+            let fromBalance := sload(fromBalanceSlot)
+            // Revert if insufficient balance.
+            if gt(amount, fromBalance) {
+                mstore(0x34, mload(0x00))
+                mstore(0x00, shl(96, 0xf6deaa04)) // `InsufficientBalance(address,uint256)`.
```
Comment (Vectorized) [inline src/tokens/ERC6909.sol:206]: Nice trick. I see what you did there. ;)

File: src/tokens/ERC6909.sol [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+    function transferFrom(address from, address to, uint256 id, uint256 amount)
+        public
+        virtual
+        returns (bool)
+    {
+        _beforeTokenTransfer(from, to, id, amount);
+        /// @solidity memory-safe-assembly
+        assembly {
+            /// Cache the free memory pointer.
+            let m := mload(0x40)
```
Comment (Vectorized) [inline src/tokens/ERC6909.sol:246]: I suspect we can also abuse the partial free-memory-pointer restore trick here. 

Gotta be careful when changing the pointers.

File: src/tokens/ERC6909.sol [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    }
+
+    /// @dev Sets `amount` as the allowance of `spender` over the caller tokens for the given `id`.
+    ///
+    /// Emits a {Approval} event.
+    function approve(address spender, uint256 id, uint256 amount) public virtual returns (bool) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            /// Cache the free memory pointer.
+            let m := mload(0x40)
```
Comment (Vectorized) [inline src/tokens/ERC6909.sol:309]: Same, I suspect we can also abuse the partial free-memory-pointer restore trick here.

File: src/tokens/ERC6909.sol [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+    event Approval(
+        address indexed owner, address indexed spender, uint256 indexed id, uint256 amount
+    );
+
+    /// @dev `keccak256(bytes("Transfer(address,address,uint256,uint256)"))`.
+    uint256 private constant _TRANSFER_EVENT_SIGNATURE =
+        0x9ed053bb818ff08b8353cd46f78db1f0799f31c9e4458fdb425c10eccd2efc44;
+
+    /// @dev `keccak256(bytes("OperatorSet(address,address,bool)"))`.
+    uint256 private constant _OPERATORSET_EVENT_SIGNATURE =
```
Comment (Vectorized) [inline src/tokens/ERC6909.sol:48]: `_OPERATOR_SET_EVENT_SIGNATURE`.

**Review by Vectorized — APPROVED** [review]
Ok, let's merge.

**Pull request merged**